quickwit-indexing/src/merge_policy.rs

```diff
@@ -204,19 +208,74 @@ impl StableMultitenantWithTimestampMergePolicy {
 
         let mut split_levels: Vec<Range<usize>> = Vec::new();
         let mut current_level_start_ord = 0;
-        let mut current_level_max_docs = (splits[0].num_records * 3).max(self.min_level_num_docs);
+        let mut current_level_max_docs =
+            (splits[0].num_records * 3).clamp(self.min_level_num_docs, self.max_merge_docs);
 
         for (split_ord, split) in splits.iter().enumerate() {
             if split.num_records >= current_level_max_docs {
                 split_levels.push(current_level_start_ord..split_ord);
                 current_level_start_ord = split_ord;
-                current_level_max_docs = 3 * split.num_records;
+                current_level_max_docs = (3 * split.num_records).min(self.max_merge_docs);
+                if split.num_records >= self.max_merge_docs {
+                    return split_levels;
+                }
             }
         }
         split_levels.push(current_level_start_ord..splits.len());
         split_levels
     }
 
+    /// Given splits tries to select a subrange of level_range that would be a good merge candidate.
+    fn merge_candidate_from_level(
+        &self,
+        splits: &[SplitMetadata],
+        level_range: Range<usize>,
+    ) -> Option<Range<usize>> {
+        let merge_candidate_end = level_range.end;
+        let mut merge_candidate_start = merge_candidate_end;
+        for split_ord in level_range.rev() {
+            if self.merge_candidate_size(&splits[merge_candidate_start..merge_candidate_end])
+                == MergeCandidateSize::OneMoreSplitWouldBeTooBig
+            {
+                break;
+            }
+            merge_candidate_start = split_ord;
+        }
+        if self.merge_candidate_size(&splits[merge_candidate_start..merge_candidate_end])
+            == MergeCandidateSize::TooSmall
+        {
+            return None;
+        }
+        Some(merge_candidate_start..merge_candidate_end)
+    }
+
+    /// Returns true iff we should stop adding extra split into this
+    /// merge candidate.
+    fn merge_candidate_size(&self, splits: &[SplitMetadata]) -> MergeCandidateSize {
```

Comment:
From what I understand we are checking if we can merge the split candidates. This depends on several conditions: number of splits, number of documents. 
I would rename some things here:


```rust
enum MergeCandidateState {
    NotEnoughSplits,
    Valid,
    TooManySplits, // we can keep too big for both
    TooManyDocuments, // 
}
```

Reply:
From the caller side, the distinction TooManySplits or  TooManyDocuments does not matter though.

I'll add explanations in the rustdoc

Reply:
oh yes, I misunderstood the usage I would have expected a condition like `if splits.len() == self.merge_factor_max` to return `MergeCandidateSize::OneMoreSplitWouldBeTooBig`. 
I guess that the case `splits.len() > self.merge_factor_max` with the current caller implementation.